pylint/typing.py

```diff
@@ -117,3 +121,14 @@ class ManagedMessage(NamedTuple):
 CheckerT_co = TypeVar("CheckerT_co", bound="BaseChecker", covariant=True)
 AstCallbackMethod = Callable[[CheckerT_co, "nodes.NodeNG"], None]
 """Callable representing a visit or leave method."""
+
+ReportsCallable = Callable[["Section", "LinterStats", Optional["LinterStats"]], None]
+"""Callable to create a report."""
+
```

Comment:
Thinking out loud here, but what if we introduce a class that does all this and make it compatible with the legacy ? (I.e. we handle both for a while but encourage the use of the class)

```python
class RawMessageDefinition(NamedTuple):

   """ This class look a lot like a ``MessageDefinition`` but without having to give the checker. """

   msgid: str #Or another class taking into account that it's a special string with a letter and a number
   symbol: str #Or another class taking into account that it's a special string with only letters and '-'
   ...
```

Reply:
Probably need to this to complete the typing of `BaseChecker` as it uses `**definition` a couple of times.

For now, I think this is fine though. Small steps towards a complete and fully typed version.